File: calc_total_jewel_burn.mjs
import { ethers } from 'ethers';
import fs from 'fs';

const ERC20_ABI = [
    {
        "constant": true,
        "inputs": [
            {
                "name": "_owner",
                "type": "address"
            }
        ],
        "name": "balanceOf",
        "outputs": [
            {
                "name": "balance",
                "type": "uint256"
            }
        ],
        "payable": false,
        "stateMutability": "view",
        "type": "function"
    }
];

async function getBalance() {
    try {

        // providers
        const HARMProvider = new ethers.JsonRpcProvider('https://api.harmony.one/');
        const DFKCProvider = new ethers.JsonRpcProvider('https://subnets.avax.network/defi-kingdoms/dfk-chain/rpc');
        const KLAYProvider = new ethers.JsonRpcProvider('https://klaytn.rpc.defikingdoms.com/');

        // total burn amounts
        const HARMJEWEL = new ethers.Contract("0x72Cb10C6bfA5624dD07Ef608027E366bd690048F", ERC20_ABI, HARMProvider);
        const DFKCJEWEL = await DFKCProvider.getBalance('0x000000000000000000000000000000000000dEaD');
        const DFKCValJEWEL = await DFKCProvider.getBalance('0x0000000000000000000000000000000000000000');
        const KLAYJEWEL = new ethers.Contract("0x30C103f8f5A3A732DFe2dCE1Cc9446f545527b43", ERC20_ABI, KLAYProvider);
        const tokenomicsBurn = 375328483.070918749502816906;

        // get and convert balances
        // harmong
        const balanceHARM = await HARMJEWEL.balanceOf('0x000000000000000000000000000000000000dEaD');
        const formattedHARM = Number(ethers.formatUnits(balanceHARM, 18));

        // DFKChain
        const formattedBalDFKCJewel = Number(ethers.formatUnits(DFKCJEWEL, 18));
        const formattedValDFKCJewel = Number(ethers.formatUnits(DFKCValJEWEL, 18));
        const totalDFKJ = formattedBalDFKCJewel + formattedValDFKCJewel - tokenomicsBurn;

        // Klaytn
        const balanceKLAY = await KLAYJEWEL.balanceOf('0x000000000000000000000000000000000000dEaD');
        const formattedKLAY = Number(ethers.formatUnits(balanceKLAY, 18));

        const totalBurn = totalDFKJ + formattedKLAY + formattedHARM;

        // Create query object
        const query = {
            day: new Date().toLocaleDateString(),
            time: new Date().toLocaleTimeString(),
            totalBurn: totalBurn + ' JEWEL',
            difference: null,
        };

        // Read previous queries from JSON file
        let previousQueries = [];
        try {
            const fileData = fs.readFileSync('queries.json');
            previousQueries = JSON.parse(fileData);
        } catch (error) {
            console.error('Error reading queries.json:', error.message);
        }

        // Calculate difference from previous query
        if (previousQueries.length > 0) {
            const previousQuery = previousQueries[previousQueries.length - 1];
            const previousTotalBurn = parseFloat(previousQuery.totalBurn);
            query.difference = totalBurn - previousTotalBurn + ' JEWEL';
        }

        // Add current query to previous queries
        previousQueries.push(query);

        // Write updated queries to JSON file
        fs.writeFileSync('queries.json', JSON.stringify(previousQueries, null, 2));

        // Display query information
        console.log('Harmony Burn:', formattedHARM);
        console.log('DFKC Burn:', totalDFKJ);
        console.log('Klaytn Burn:', formattedKLAY);
        console.log('\nTotal Burn:', totalBurn + ' JEWEL');
        console.log('Difference from previous query:', query.difference);

    } catch (error) {

        console.error('Error:', error.message);
    }
}

getBalance();

File: calc_total_jewel_burn.py
from web3 import Web3
import json
import os
import datetime

# ABI for ERC20 token
ERC20_ABI = [
    {
        "constant": True,
        "inputs": [{"name": "_owner", "type": "address"}],
        "name": "balanceOf",
        "outputs": [{"name": "balance", "type": "uint256"}],
        "payable": False,
        "stateMutability": "view",
        "type": "function"
    }
]

# Function to get balances
async def get_balance():
    try:
        # Providers
        HARMProvider = Web3(Web3.HTTPProvider('https://api.harmony.one/'))
        DFKCProvider = Web3(Web3.HTTPProvider('https://subnets.avax.network/defi-kingdoms/dfk-chain/rpc'))
        KLAYProvider = Web3(Web3.HTTPProvider('https://klaytn.rpc.defikingdoms.com/'))

        # Contract instances
        HARMJEWEL = HARMProvider.eth.contract(address="0x72Cb10C6bfA5624dD07Ef608027E366bd690048F", abi=ERC20_ABI)
        KLAYJEWEL = KLAYProvider.eth.contract(address="0x30C103f8f5A3A732DFe2dCE1Cc9446f545527b43", abi=ERC20_ABI)

        # Total burn amounts
        tokenomicsBurn = 375328483.070918749502816906

        # Get and convert balances
        balanceHARM = HARMJEWEL.functions.balanceOf('0x000000000000000000000000000000000000dEaD').call()
        formattedHARM = balanceHARM / 1e18

        DFKCJEWEL = DFKCProvider.eth.get_balance('0x000000000000000000000000000000000000dEaD') / 1e18
        DFKCValJEWEL = DFKCProvider.eth.get_balance('0x0000000000000000000000000000000000000000') / 1e18
        totalDFKJ = DFKCJEWEL + DFKCValJEWEL - tokenomicsBurn

        balanceKLAY = KLAYJEWEL.functions.balanceOf('0x000000000000000000000000000000000000dEaD').call()
        formattedKLAY = balanceKLAY / 1e18

        totalBurn = totalDFKJ + formattedKLAY + formattedHARM

        # Create query object
        query = {
            "day": str(datetime.date.today()),
            "time": str(datetime.datetime.now().time()),
            "totalBurn": f"{totalBurn} JEWEL",
            "difference": None
        }

        # Read previous queries from JSON file
        previousQueries = []
        if os.path.exists('queries.json'):
            with open('queries.json', 'r') as file:
                previousQueries = json.load(file)

        # Calculate difference from previous query
        if previousQueries:
            previousTotalBurn = float(previousQueries[-1]["totalBurn"].split()[0])
            query["difference"] = f"{totalBurn - previousTotalBurn} JEWEL"

        # Add current query to previous queries
        previousQueries.append(query)

        # Write updated queries to JSON file
        with open('queries.json', 'w') as file:
            json.dump(previousQueries, file, indent=2)

        # Display query information
        print(f"Harmony Burn: {formattedHARM}")
        print(f"DFKC Burn: {totalDFKJ}")
        print(f"Klaytn Burn: {formattedKLAY}")
        print(f"\nTotal Burn: {totalBurn} JEWEL")
        print(f"Difference from previous query: {query['difference']}")

    except Exception as e:
        print(f"Error: {e}")

# Run the function
import asyncio
asyncio.run(get_balance())
